Fix uncaught gate exception on data migrations admin calls

bashtanov opened the pull request:
See [here, lines 5569-5587](https://ci-artifacts.dev.vectorized.cloud/redpanda/72238/01994bdd-b537-4b62-984a-f41860903110/vbuild/ducktape/results/2025-09-15--001/DataMigrationsApiTest/test_mount_inexistent/188/RedpandaService-0-281471214933792/docker-rp-33/redpanda.log) (search for `06:26:59,152`).

Raft0 shutdown interferes with an admin api request. Normal `mux_state_machine::replicate_and_wait` calls nicely report `errc::shutting_down` in case of a relevant shutdown exception. Do the same for `controller_stm::insert_linearizable_barrier`.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [x] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
c/data-migrate/frontend: simplify linearizable barrier function

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--9e85ebbf-d906-4133-863e-8ccc450b1af9-->
<!--TEST RESULTS-->
<!--957b3f38-0785-4eda-8896-944b8dbd75b8-->
<!--TEST RESULTS-->
<!--1654292a-e139-4e4f-911c-c7baf9fbc5da-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73457">build#73457</a></summary>


| test_class           | test_method          | test_arguments                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     | test_history                                                                                                                                            |
|:---------------------|:---------------------|:--------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi | test_no_auth_enabled | {"audit_transport_mode": "rpc"} | integration | https://buildkite.com/redpanda/redpanda/builds/73457#0199a565-2f76-43ac-a260-0aac82d4d832 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73551">build#73551</a></summary>


| test_class                      | test_method                 | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      | test_history                                                                                                                                                              |
|:--------------------------------|:----------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests            | test_delete_simple_link     | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c470-4d49-a582-bd586388d5b2 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_delete_simple_link                |
| ShadowLinkBasicTests            | test_delete_simple_link     | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c48f-4458-b400-06274602afaf | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_delete_simple_link                |
| ShadowLinkTopicFailoverTests    | test_link_failover          | {"with_failures": false}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c466-4c7b-86ab-ee9b7e32d1ea | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_failover             |
| ShadowLinkTopicFailoverTests    | test_link_failover          | {"with_failures": false}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c483-4a60-b632-717c0d6a6992 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_failover             |
| ShadowLinkTopicFailoverTests    | test_link_failover          | {"with_failures": true}                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c468-43e2-9038-305b8b97c5d0 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_failover             |
| ShadowLinkTopicFailoverTests    | test_link_failover          | {"with_failures": true}                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c485-4d9b-9c07-5bc8de478a55 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_failover             |
| ShadowLinkTopicFailoverTests    | test_link_topic_failover    | {"with_failures": false}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c469-415f-8f45-04afe8d22c46 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_topic_failover       |
| ShadowLinkTopicFailoverTests    | test_link_topic_failover    | {"with_failures": false}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c487-4410-af21-a53c35a72126 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_topic_failover       |
| ShadowLinkTopicFailoverTests    | test_link_topic_failover    | {"with_failures": true}                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c46b-4db9-9df4-81f47d35c3d5 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_topic_failover       |
| ShadowLinkTopicFailoverTests    | test_link_topic_failover    | {"with_failures": true}                                                                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c489-4144-9bb4-9d7b2dbca45e | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkTopicFailoverTests&test_method=test_link_topic_failover       |
| AdminV2ListKafkaConnectionsTest | test_list_kafka_connections | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c46b-4db9-9df4-81f47d35c3d5 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AdminV2ListKafkaConnectionsTest&test_method=test_list_kafka_connections |
| AdminV2ListKafkaConnectionsTest | test_list_kafka_connections | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c489-4144-9bb4-9d7b2dbca45e | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AdminV2ListKafkaConnectionsTest&test_method=test_list_kafka_connections |
| RandomNodeOperationsTest        | test_node_operations        | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73551#0199aae0-c493-468d-8407-e72f3b353476 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76552">build#76552</a></summary>


| test_class                        | test_method                                  | test_arguments                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                               |
|:----------------------------------|:---------------------------------------------|:----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTxRemovalUpgradeTest | test_tx_control_batch_removal_with_upgrade   | {"test_case_name": "All commits"} | integration | https://buildkite.com/redpanda/redpanda/builds/76552#019a9770-83f9-4f23-9a98-281462fe23fe | FLAKY         | 10/21    | upstream reliability is '88.78504672897196'. current run reliability is '47.61904761904761'. drift is 41.166 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
| SegmentMsTest                     | test_segment_rolling_with_retention_consumer | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/76552#019a9770-83f4-49fb-97db-85514a4d7706 | FLAKY         | 19/21    | upstream reliability is '91.96927374301676'. current run reliability is '90.47619047619048'. drift is 1.49308 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer                   |
| RedpandaMixedTestSelfTest         | test_metrics                                 | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/76552#019a977a-1af8-4dc0-ba49-59ffc0ab82ed | FLAKY         | 11/21    | upstream reliability is '79.14012738853503'. current run reliability is '52.38095238095239'. drift is 26.75918 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaMixedTestSelfTest&test_method=test_metrics                                       |
</details>
<!--/body-->

dotnwat commented on src/v/cluster/controller_stm.cc:
```diff
@@ -239,4 +239,16 @@ ss::future<std::error_code> controller_stm::replicate_and_wait(
       });
 }
 
+ss::future<result<model::offset>> controller_stm::insert_linearizable_barrier(
```
i'm guessing the commits should be reordered, otherwise the first won't compile?

bashtanov replied:
why? the first one is just to reduce code duplication

dotnwat replied:
i think you read/answered this question after your force push

joe-redpanda commented on src/v/cluster/controller_stm.h:
```diff
@@ -87,6 +87,9 @@ class controller_stm final
       std::optional<model::term_id> term = std::nullopt);
 
+    ss::future<result<model::offset>>
+    insert_linearizable_barrier(model::timeout_clock::time_point timeout);
```
any reason to not push this up to state_machine.cc

bashtanov replied:
Originally my idea was to avoid returning `raft::errc::shutting_down` for state machines attached to data partitions, so that we don't return `raft::errc::shutting_down`, as I thought it might be treated as app shutdown by handlers down the stack. But it looks it's handled correctly pretty much everywhere, the worst I can see is [this log message](https://github.com/redpanda-data/redpanda/blob/c245324a83472d3d00856752d7ceb3fd7a713340/src/v/kafka/server/group.cc#L2723). Let me change it and run the tests.

bashtanov committed:
r/state_machine: linearizable_barrier to return raft::shutdown not throw

Shutdown errors happen on STM or raft shutdown, return relevant error
code instead of throwing.

bashtanov commented:
/ci-repeat 1
skip-redpanda-build
skip-unit

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73551-->
<!--d8208737-3254-4349-ad1b-d9e5030786a8-->
<!--40ff8c9b-bda8-4c9e-9138-8d664e9c4a93-->
<!--512267bc-cbd1-4083-9fa6-d363952f5e22-->
<!--445e0ef2-73c9-499f-89dc-f371768c4f65-->
<!--e6092160-a7c5-4eb7-b864-68e3cfb6fc52-->
<!--37b17787-7e56-48b8-8abd-0d4ef6b0faee-->
<!--9c8efe37-105b-4723-be18-6072619f8e68-->
<!--801ac3fb-1e01-4455-9743-3b12467928a6-->
<!--937b3ba1-4610-4efd-88cb-0f6637b8c482-->
<!--eed2d126-2711-4b2b-8ce7-622e53274d30-->
<!--title-->
## Retry command for [Build#73551](https://buildkite.com/redpanda/redpanda/builds/73551)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/list_kafka_connections_test.py::AdminV2ListKafkaConnectionsTest.test_list_kafka_connections
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkTopicFailoverTests.test_link_topic_failover@{"with_failures":true}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkBasicTests.test_delete_simple_link
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkTopicFailoverTests.test_link_failover@{"with_failures":false}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkTopicFailoverTests.test_link_failover@{"with_failures":true}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkTopicFailoverTests.test_link_topic_failover@{"with_failures":false}
```
<!--/body-->

bharathv commented on src/v/raft/state_machine.cc:
```diff
@@ -207,7 +207,14 @@ ss::future<result<model::offset>> state_machine::insert_linearizable_barrier(
+          }
+          std::rethrow_exception(e);
       });
+    ;
```
extra ;?

bharathv commented on src/v/raft/state_machine.cc:
```diff
@@ -207,7 +207,14 @@ ss::future<result<model::offset>> state_machine::insert_linearizable_barrier(
+          if (ssx::is_shutdown_exception(e)) {
+              return result<model::offset>(errc::shutting_down);
+          }
+          std::rethrow_exception(e);
```
shouldn't we return an exceptional future here instead of rethrowing it?

dotnwat replied:
> shouldn't we return an exceptional future here instead of rethrowing it?

@bharathv yes we should. technically handle_exception should do that conversion for us, but it's added overhead and it's hard to know from one continuation to the next which ones do the conversion and which don't.

dotnwat reviewed (dismissed):
avoid throwing in continuation. see my/@bharath's review comment

bashtanov committed:
r/state_machine: simplify insert_linearizable_barrier

by using a type alias for its return type

bharathv commented on src/v/raft/state_machine.cc:
```diff
@@ -198,8 +198,8 @@ state_machine::insert_linearizable_barrier(
 
-    return _raft->linearizable_barrier(timeout).then(
-      [this, timeout](result<model::offset> r) {
+    return _raft->linearizable_barrier(timeout)
```
unrelated to this PR, wondering if this method should hold the stm gate.

bashtanov replied:
We lack this in pretty much all methods of all STMs. I have it on my list to fix.

bashtanov commented:
@dotnwat did I address your concerns?
